anesthetic/utils.py

```diff
@@ -89,6 +90,151 @@ def quantile(a, q, w=None, interpolation='linear'):
     return quant
 
 
+def sample_cdf(samples, inverse=False, interpolation='linear'):
+    """Sample the empirical cdf for a 1d array."""
+    samples = np.sort(samples)
+    ngaps = len(samples)-1
+    gaps = np.random.dirichlet(np.ones(ngaps))
+    cdf = np.array([0, *np.cumsum(gaps)])
+    assert np.isclose(cdf[-1], 1, atol=1e-9, rtol=1e-9), \
+        "Error: CDF does not reach 1 but "+str(cdf[-1])
+    # Set the last element (tested to be approx 1)
+    # to exactly 1 to avoid interpolation errors
+    cdf[-1] = 1
+    if inverse:
+        return interp1d(cdf, samples, kind=interpolation)
+    else:
+        return interp1d(samples, cdf, kind=interpolation,
+                        fill_value=(0, 1), bounds_error=False)
+
+
+def credibility_interval(samples, weights=None, level=0.68, method="iso-pdf",
+                         return_covariance=False, nsamples=12,
+                         verbose=False):
+    """Compute the credibility interval of weighted samples.
+
+    Based on linear interpolation of the cumulative density function, thus
+    expect discretization errors on the scale of distances between samples.
+
+    https://github.com/Stefan-Heimersheim/fastCI#readme
+
+    Parameters
+    ----------
+    samples : array
+        Samples to compute the credibility interval of.
+    weights : array, default=np.ones_like(samples)
+        Weights corresponding to samples.
+    level : float, default=0.68
+        Credibility level (probability, <1).
+    method : str, default='iso-pdf'
+        Which definition of interval to use:
+
+        * ``'iso-pdf'``: Calculate iso probability density interval with the
+          same probability density at each end. Also known as
+          waterline-interval or highest average posterior density interval.
+          This is only accurate if the distribution is sufficiently unimodal.
+        * ``'lower-limit'``/``'upper-limit'``: Lower/upper limit. One-sided
+          limits for which ``level`` fraction of the (equally weighted) samples
+          lie above/below the limit.
+        * ``'equal-tailed'``: Equal-tailed interval with the same fraction of
+          (equally weighted) samples below and above the interval region.
+
+    return_covariance: bool, default=False
+        Return the covariance of the sampled limits, in addition to the mean
+    nsamples : int, default=12
+        Number of CDF samples to improve `mean` and `std` estimate.
```

Comment:
Intuitively, I expect other uncertainties (finite amount of samples) to outweigh the `nsamples=12` error, and this seems to be the case for the two examples below. By default `nsamples=1` was fine, we just added a few more because it was a very low-hanging fruit.

I'm less certain about the covariance returned. But I also have no idea how many `nsamples` are good, I don't want to make this needlessly slow.

Here's `normal_samples` for different `nsamples` values
```
1 samples: [1.90273801 2.09455387]
2 samples: [1.90269773 2.09849627]
4 samples: [1.89775164 2.09207853]
8 samples: [1.90027277 2.09531987]
12 samples: [1.89883413 2.09379086]
120 samples: [1.89997834 2.09506267]
1200 samples: [1.89981016 2.09489933]
```

Here's ./tests/example_data/pc "x0"]
```
1 samples: [-0.1988116   0.18185223]
2 samples: [-0.17449234  0.1926399 ]
4 samples: [-0.17046529  0.17364958]
8 samples: [-0.18633088  0.17586416]
12 samples: [-0.19075288  0.17333953]
120 samples: [-0.18451308  0.17946344]
1200 samples: [-0.18678313  0.1744095 ]
```